### PR TITLE
ScrollablePane and Sticky: New components and DetailsList locked header feature

### PR DESCRIPTION
**Pull request by leddie24**
#### Pull request checklist

- [x] Include a change request file using `$ npm run change`

#### Description of changes

- Adds ScrollablePane and Sticky component
- Adds locked header feature for details list

Basic example (with colored background divs)
screen shot 2017-07-19 at 3 54 17 pm

DetailsList locked header

screen shot 2017-07-19 at 3 54 25 pm

#### Focus areas to test

DetailsList, DetailsHeader, Sticky, ScrollablePane

### CODE REVIEW

**Comment by betrue-final-final** [comment]
I installed your package, but I'm not finding the example or behavior in DetailsList.sample page. Is it somewhere else?

**Comment by leddie24** [comment]
@betrue-final-final It's on the ScrollablePane page.

**Comment by betrue-final-final** [comment]
Isn't it on the detailslist though?

**Comment by betrue-final-final** [comment]
Ok, does not seem to be locking. Column header just scrolls. Panes stop about 80px from the top then disappear. I'm on Edge.
image

**Comment by leddie24** [comment]
@betrue-final-final thanks for catching that; I didn't test in Edge; it should be working properly now.

**Comment by betrue-final-final** [comment]
Alright, the colored pane version looks good. However the Detailslist version is losing the column headers. We want this to lock as well.
image

**Comment by betrue-final-final** [comment]
Took a look, and it looks great on Chrome. I'm afraid Edge being jumpy may be an Edge problem, or Thomas knows a trick.

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.Props.ts [hunk 1]
```diff
@@ -186,6 +186,9 @@ export interface IDetailsListProps extends React.Props<DetailsList> {
 
   /** Boolean value to indicate if the component should render in compact mode. Set to false by default */
   compact?: boolean;
+
+  /** Boolean value to indicate if the DetailsList Header should render in a Sticky component. Set to false by default */
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.Props.ts]: I think the best thing to do here would be to force the caller to use `onRenderHeader` and wrap the original `defaultRender` with a `Sticky` component. Don't force `DetailsList` to take a direct dependency on `Sticky`.

File: packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import * as stylesImport from './ScrollablePane.scss';
+const styles: any = stylesImport;
+
+export class ScrollablePane extends BaseComponent<IScrollablePaneProps, {}> {
+  public static childContextTypes = {
+    subscribe: PropTypes.func,
+    addStickyHeader: PropTypes.func,
+    removeStickyHeader: PropTypes.func,
+    addStickyFooter: PropTypes.func,
+    removeStickyFooter: PropTypes.func
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx]: Collapse all of these into a single `scrollablePane` object with a defined interface.

File: packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      });
+      setTimeout(() => {
+        this._resizeContainer();
+        this._stickyContainer.parentElement.removeChild(this._stickyContainer);
+        this._scrollElement.parentElement.insertBefore(this._stickyContainer, this._scrollElement.nextSibling);
+        this._notifySubscribers();
+        this._sortFooters();
+      }, 500);
+    } else {
+      throw new TypeError('Expected ScrollablePane to be within a parent element with data-is-scrollable=true');
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx]: To me, `ScrollablePane` should supply the element with `data-is-scrollable="true"`, not depend on it. It should also supply the standard `overflow` and `-webkit-overflow-scroll: touch` behaviors.

File: packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    const { className } = this.props;
+
+    return (
+      <div className={ css('ms-ScrollablePane', className) }
+        ref='root'>
+        <div ref='stickyContainer' className={ styles.stickyContainer }>
+          <div ref='stickyAbove' className={ styles.stickyAbove } />
+          <div ref='stickyBelow' className={ styles.stickyBelow } />
+        </div>
+        { this.props.children }
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx]: Should wrap the children with the `<div data-is-scrollable="true" />` element.

File: packages/office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+          <Sticky>
+            <h1 style={ { margin: '0px' } }>Item List</h1>
+          </Sticky>
+          <MarqueeSelection selection={ this._selection }>
+            <DetailsList
+              items={ items }
+              columns={ _columns }
+              setKey='set'
+              layoutMode={ DetailsListLayoutMode.fixedColumns }
+              isLockedHeader={ true }
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx]: To supply `<Sticky>`, a consumer should do:
```tsx
onRenderDetailsHeader={
    (detailsHeaderProps: IDetailsHeaderProps, defaultRender: IRenderFunction<IDetailsHeaderProps>) => (
        <Sticky>
            { defaultRender({
                ...detailsHeaderProps,
                onRenderColumnHeaderTooltio: (tooltipHostProps: ITooltipHostProps) => <TooltipHost { ...tooltipHostProps } />
            }) }
        </Sticky>
    ) }
```

File: packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+const styles: any = stylesImport;
+
+export interface IStickyState {
+  isStickyTop: boolean;
+  isStickyBottom: boolean;
+  placeholderHeight?: number;
+}
+
+export class Sticky extends BaseComponent<IStickyProps, IStickyState> {
+  public static contextTypes = {
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx]: Group into a single field with an interface.

File: packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx (hunk 2 above)
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx]: Hey @ThomasMichon , made changes here, but not 100% sure if this is correct; can you check my work? Thanks

File: packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx (hunk 4 above)
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx]: data-is-scrollable="true" is on the root div

File: packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.Props.ts (hunk 1 above)
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.Props.ts]: Fixed, removed isLocked prop and changed to onRenderHeader

File: packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx (hunk 3 above)
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/ScrollablePane.tsx]: Done, changed to have ScrollablePane be scrollable itself.  I set the max-height: inherit on the element so it fills the parent space

File: packages/office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx (hunk 5 above)
Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/ScrollablePane/examples/ScrollablePane.DetailsList.Example.tsx]: Fixed, thanks

**Comment by leddie24** [comment]
This is ready for review again @ThomasMichon.  I'm not sure on interfacing context with Typescript, can I get your assistance on that?  Thanks!

File: packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    root: HTMLElement;
+    placeholder: HTMLElement;
+  };
+
+  public context: {
+    subscribe: (handler: Function) => void;
+    addStickyHeader: (sticky: Sticky) => void;
+    removeStickyHeader: (sticky: Sticky) => void;
+    addStickyFooter: (sticky: Sticky) => void;
+    removeStickyFooter: (sticky: Sticky) => void;
```
Comment (ThomasMichon) [inline packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx]: These should not be individual fields under `context`. This means that anyone would collide on a field named `subscribe`. You should make it look this this:

```ts
public context: {
    scrollablePane: IScrollablePane;
}
```

Comment (leddie24) [inline packages/office-ui-fabric-react/src/components/Sticky/Sticky.tsx]: Thanks, can you re-review when you have a moment?

**Pull request merged**